sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -148,3 +148,111 @@ private[joins] object HashedRelation {
     }
   }
 }
+
+
+/**
+ * A HashedRelation for UnsafeRow, which is backed by BytesToBytesMap that maps the key into a
+ * sequence of values.
+ *
+ * TODO(davies): use BytesToBytesMap
+ */
+private[joins] final class UnsafeHashedRelation(
+    private var hashTable: JavaHashMap[UnsafeRow, CompactBuffer[UnsafeRow]],
+    private var keyTypes: Array[DataType],
+    private var rowTypes: Array[DataType])
+  extends HashedRelation with Externalizable {
+
+  def this() = this(null, null, null)  // Needed for serialization
+
+  // UnsafeProjection is not thread safe
+  @transient lazy val keyProjection = new ThreadLocal[UnsafeProjection]
+  @transient lazy val fromUnsafeProjection = new ThreadLocal[FromUnsafeProjection]
+
+  override def get(key: InternalRow): CompactBuffer[InternalRow] = {
+    val unsafeKey = if (key.isInstanceOf[UnsafeRow]) {
+      key.asInstanceOf[UnsafeRow]
+    } else {
+      var proj = keyProjection.get()
+      if (proj eq null) {
+        proj = UnsafeProjection.create(keyTypes)
+        keyProjection.set(proj)
+      }
+      proj(key)
+    }
+
+    val values = hashTable.get(unsafeKey)
+    // Return GenericInternalRow to work with other JoinRow, which
+    // TODO(davies): return UnsafeRow once we have UnsafeJoinRow.
```

Comment:
If we're not going to implement this as part of this patch, then let's make sure to file a followup JIRA under the Tungsten umbrella.